README.md

```diff
@@ -385,6 +385,39 @@ foreman-maintain console
 foreman-maintain config
 ```
 
+## Bash completion
+
+Bash completion is automatically installed by RPM. To use it for development setup 
+`cp ./config/foreman-maintain.completion /etc/bash_completion.d/foreman-maintain` 
+and load it to the current shell `source /etc/bash_completion.d/foreman-maintain`.
+
+Bash completion for foreman-maintain needs pre-built cache that holds description of 
+all subcommands and its parameters. The cache is located by default in 
+`~/.cache/foreman_maintain_completion.yml`. The location can be changed in foreman-maintain's
+config file. The cache is pre-built automatiacally during RPM install, can be built manually
+with `foreman-maintain advanced prebuild-bash-completion` or is built automatically when
+completion is used and the cache is missing (this may cause slight delay).
+
+The completion offers suggestion of possible command-line subcommands and their options
+as usual. It can also suggest values for options and params where it should be file
+or directory path.
+
+####  Available value types
+
+Completion of values is dependent on CLI option and prameter settings, e.g.:
+
+```ruby
+  parameter 'BACKUP_DIR', 'Path to backup dir', :completion => { :directory => {} }
+```
+
+Possible options for the `:completion` attribute are:
+ - `{}` option has no value, default for flags
+ - `{ :value => {} }` option has value of unknown type, no suggestions for the value, default
```

Comment:
The empty hash in `{ :value => {} }` looks a bit odd. What about somehing like: `{ :type => :value }`, for file it could be `{ :type => :file, :filter => '\.txt$'}`

Reply:
Or, for simpler cases, maybe `{ :completion => :directory }`

Reply:
One suggestion - doesn't have to be solved in this PR, but would be good to think about this now: It would be great if `foreman-maintain upgrade check --target-version ` could produce a list of versions that the system can be upgraded to: since this list could be dynamic, we would need to generate it on actual setup (rather than in packaging time). I can imagine the setting by something like `{ :type => :value, :generator => lambda { Upgrades.available_versions } }` - warping to lambda so that it would not be evaluated every time the file is sourced.

Reply:
@inecas I'll re-check on notation, what you suggest  feels more natural.
Regarding the target version value, I have some plans for this already. The cache is never build during packaging but rather during rpm %post. Then we can have 
```ruby
{ :type => :enum, :values => Upgrades.available_versions } 
```
We should probably call the cache rebuild as part of upgrade scenario.

The lambda probably won't work as we don't load whole ForemanMaintain (for performance reasons) just the completion parser, so no FM objects are available during the parsing.

Part of more ambitious plans is to add `:function` type. In the project you would inherit from the `Completion` class and implement there methods returning the more dynamic data on the fly e.g. `organizations', `products`, `capsules`, etc.. In the option you could do
```ruby
{ :type => :function, :name => :capsules }
```

Reply:
> The lambda probably won't work as we don't load whole ForemanMaintain

The idea is that the lambda would be called only on cache-creation phase, to store the values in the cache, not in the completion parser.